### DISCUSSION TITLE
How do I access embedded etcd when running K3S on armhf

### DISCUSSION THREAD

**Discussion by daveneeley** [k3s-io/k3s#8968] [Q&A] [answered]
https://github.com/k3s-io/k3s/issues/2732 shows some patterns for connecting to etcd running within the k3s binary by using an external etcd docker container. Unfortunately the nodes running my kube/etcd masters are 32 bit ARM (armhf) and after checking a few places I've not found anyone who compiles containers for it.

Is there another route I can take to access etcd within the K3S binary on armhf?

**Comment by brandond** [comment]
Are you sure you're using etcd? By default k3s uses kine+sqlite instead of embedded etcd, and we don't generally recommend etcd on devices that use SD cards for storage.

What specifically are you trying to do?

**Reply by daveneeley** [reply]
Similar to the linked issue, I had attempted to delete a master node with `kubectl delete node`. Then I used the uninstall script to clean that server. Then I reinstalled the master. Since it was the same host it had the same node name. However when the node tried to join as a master, there was an error message about that node name already existing, despite my deleting it.

My assumption was that evidence of the master node needed to be deleted from etcd so it could be re-added. It sound alike it needed to be deleted from sqllite instead.

As far as armhf, these are Hardkernel Odroid small board computers with eMMC "disk". They aren't subject to the SD card issues.

**Reply by brandond** [reply, marked as answer]
> However when the node tried to join as a master, there was an error message about that node name already existing, despite my deleting it.
> My assumption was that evidence of the master node needed to be deleted from etcd so it could be re-added. It sound alike it needed to be deleted from sqllite instead.

No, poking at etcd is not the correct thing to do, and won't get you anywhere. Check for the node config secret documented here: https://docs.k3s.io/architecture?_highlight=password%20secret#how-agent-node-registration-works

If deleting that secret does not allow the node to rejoin, restart the existing server and try again - I have seen an occasional unreproducible reports of the node rejoins being blocked even after the secret is removed, that seem to be resolved by a restart of the service.

**Reply by daveneeley** [reply]
I think I used `--cluster-reset` but this looks better. Thanks for your help!

**Reply by brandond** [reply]
--cluster-reset removes other nodes from the etcd cluster, but won't remove anything from Kubernetes - which is what would be blocking you from rejoining if you see an error about the node password.

**Reply by daveneeley** [reply]
I did this
- ran uninstall script on two of three server nodes (the one that couldn't rejoin + another one), this was the second time uninstalling the one on the host that couldn't rejoin
- checked that nodes were deleted with kubectl get nodes (remember I had already deleted the one that could not rejoin with kubectl delete, so this was just checking for the other one)
- restarted remaining node with --cluster-reset
- restarted again without --cluster-reset, maybe? It was a few nights ago and it was late, sorry
- re-installed other two server nodes

When restarting the cluster the same server join token was used. Sounds like, since I'm not using etcd, that --cluster-reset should have done nothing? Perhaps the restart of the remaining server node was what actually addressed the issue.

Definitely what I did was overkill.
